drivers/sqliter-driver/src/commonMain/kotlin/com/squareup/sqldelight/sqliter/driver/SQLiterHelper.kt

```diff
@@ -0,0 +1,228 @@
+ * in a "Saner Concurrency" context, we'd be better off if connections were tied to threads.
+ *
+ * SQLite cannot nest transactions, so there should be no situation where we have an enclosing transaction.
+ * Transactions are isolated to connections, though, so if you're looking to have multiple transactions,
+ * create multiple connections. This won't help a lot from a performance perspective because only one
+ * connection can write at one time.
+ */
+class SQLiterConnection(
+        internal val databaseConnection: DatabaseConnection
+) : SqlDatabaseConnection, EnforceClosed by EnforceClosedImpl() {
+    private val transaction: AtomicReference<Transaction?> = AtomicReference(null)
+    private val transLock = QuickLock()
+    private val statementList = frozenLinkedList<Statement>(stableIterator = false)
+    private val queryList = frozenLinkedList<SQLiterQuery>(stableIterator = false)
+
+    override fun currentTransaction(): Transacter.Transaction? = transaction.value
+
+    override fun newTransaction(): Transacter.Transaction =
+            transLock.withLock {
+                if (transaction.value != null)
+                    throw IllegalStateException("Transaction already active")
+                databaseConnection.beginTransaction()
+                val trans = Transaction()
+                transaction.value = trans
+                return trans
+            }
+
+    override fun prepareStatement(sql: String, type: SqlPreparedStatement.Type, parameters: Int): SqlPreparedStatement {
+        checkNotClosed()
+        val statment = databaseConnection.createStatement(sql)
+        statementList.add(statment)
+        return when(type) {
+            SELECT -> {
+                val query = SQLiterQuery(sql, databaseConnection)
+                queryList.add(query)
+                query
+            }
+            INSERT, UPDATE, DELETE, EXECUTE -> SQLiterStatement(statment, type)
+        }
+    }
+
+    internal fun close(){
+        checkNotClosed()
+        trackClosed()
+        statementList.forEach { it.finalizeStatement() }
+        queryList.forEach { it.close() }
+        databaseConnection.close()
+    }
+
+    inner class Transaction : Transacter.Transaction() {
+        override val enclosingTransaction: Transacter.Transaction? = null
+
+        override fun endTransaction(successful: Boolean) = transLock.withLock {
+            if (successful) {
+                databaseConnection.setTransactionSuccessful()
+                databaseConnection.endTransaction()
+            } else {
+                databaseConnection.endTransaction()
+            }
+
+            transaction.value = null
+        }
+    }
+}
+
+class SQLiterQuery(private val sql: String,
+                   private val database: DatabaseConnection) :
+        SqlPreparedStatement, EnforceClosed by EnforceClosedImpl() {
+    internal val statementList = frozenLinkedList<Statement>(stableIterator = false)
```

Comment:
Each "cursor" ultimately points to a sqlite statement. Multiple open cursors need separate sqlite statement instances, which need to be tracked, reused, and closed before closing the connection.